serdes-native/src/main/java/io/ballerina/stdlib/serdes/Deserializer.java

```
@@ -0,0 +1,315 @@
+            ArrayType arrayType = (ArrayType) type;
+            Type elementType = arrayType.getElementType();
+            FieldDescriptor fieldDescriptor = schema.findFieldByName(ARRAY_FIELD_NAME);
+            schema = fieldDescriptor.getContainingType();
+            return getBallerinaArrayValueFromMessage(dynamicMessage.getField(fieldDescriptor), elementType, schema, 1);
+        } else if (type.getTag() == TypeTags.RECORD_TYPE_TAG) {
+            Map<String, Object> mapObject = getBallerinaRecordValueFromMessage(dynamicMessage, type, schema);
+            return ValueCreator.createRecordValue(type.getPackage(), type.getName(), mapObject);
+        } else {
+            throw createSerdesError(UNSUPPORTED_DATA_TYPE + type.getName(), SERDES_ERROR);
+        }
+    }
+
+    private static Object getBallerinaPrimitiveValueFromMessage(Object value) {
+        String valueInString = value.toString();
+
+        switch (value.getClass().getSimpleName()) {
+            case STRING:
+                return StringUtils.fromString(valueInString);
+            case DOUBLE:
+                return ValueCreator.createDecimalValue(valueInString);
+            default:
+                return value;
+        }
+    }
+
+    private static Object getBallerinaArrayValueFromMessage(Object value, Type type, Descriptor schema,
+                                                            int unionFieldIdentifier) {
+        if (value.getClass().getSimpleName().equals(BYTE)) {
+            ByteString byteString = (ByteString) value;
+            return ValueCreator.createArrayValue(byteString.toByteArray());
+        } else {
+            Collection collection = (Collection) value;
+            BArray bArray = ValueCreator.createArrayValue(TypeCreator.createArrayType(type));
+            for (Object element : collection) {
+                if (type.getTag() == TypeTags.STRING_TAG) {
+                    bArray.append(StringUtils.fromString(element.toString()));
```

Using `toString()` should be avoided. The better way to do this is to cast the value.

If we check the type tag, the value must be the type we are expecting. Then we should be able to cast the `element` value to the exact Java type. If the cast cannot be done, it should throw an error. 

Using `toString()` can lead to incorrect behavior.

---

Fixed